### ISSUE TITLE
Allow for chaining DataType.asConvertedDataType(Converter) calls

### ISSUE THREAD

**Issue by victorbr** [jOOQ/jOOQ#8076]
### Expected behavior and actual behavior:

When calling `asConvertedDataType` on a `DataType` produced with `asConvertedDataType`, the conversion functions are not composed as one would expect.

The example is surely not a real-life one, just to demonstrate the problem. A `VARCHAR` field is first converted to `Long` while replacing all non-digit characters by 0 and then converted to `Integer`.
A value `123q456` is inserted into the database, to validate that it passes through the first `Converter`. In reality jOOQ tries to convert `String` to `Long` automatically and fails.

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

https://github.com/victorbr/jOOQ-mcve/tree/victorb/asConvertedDataType

### Versions:

- jOOQ:
- Java:
- Database (include vendor):
- OS:
- JDBC Driver (include name if inofficial driver):

P.S. `jOOQ-mvce` is awesome!!

**Comment by lukaseder** [comment]
> P.S. jOOQ-mvce is awesome!!

Yeah, isn't it? :-)

Thanks a lot for your report. Will look into it right away.

**Comment by lukaseder** [comment]
Hmm, indeed - I can see how this might be the expectation. Of course, you can easily chain your `Converter` instances in your case by using `Converters.of(converter1, converter2)`. With converters, the chaining seems obvious, but what about when using bindings? Both bindings "think" they are the ones binding user values `<U>` to JDBC as `<T>` (or as some other type than `<T>`). Which one should apply?

However, I do think that deriving a data type from an existing type using converters should work, if no additional bindings are involved. Your example makes perfect sense.

**Comment by lukaseder** [comment]
Another test that could be expected to work, but currently doesn't:

```java
    @Test
    public void testChainedConvertersOnExistingDataTypes() {
        Converter<String, Long> c1 =
            Converter.ofNullable(String.class, Long.class, Long::parseLong, Object::toString);
        Converter<Long, Integer> c2 =
            Converter.ofNullable(Long.class, Integer.class, Number::intValue, Number::longValue);

        DataType<Long> d1 = SQLDataType.VARCHAR.asConvertedDataType(c1);
        DataType<Integer> d2 = d1.asConvertedDataType(c2);

        assertEquals(1, (int) ((Converter<String, Integer>) d2.getConverter()).from("1"));
        assertEquals("1", d2.getConverter().to(1));
    }
```

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.12. I'm a bit reluctant to backport this change to 3.11 and older releases. It kinda feels like a breaking change, even if the new behaviour is better. Will leave this in 3.12 only.

**Comment by victorbr** [comment]
Lukas, thank you so much! A prompt and to-the-point response as usual! I'll consider using the workaround of `Converters.of` until 3.12 is released.
